contributors/design-proposals/code-generator/kubegen.md

```diff
@@ -0,0 +1,333 @@
+- prepend LICENSE to generated files (wrapping in comments)
+
+### Options
+
+A number of options can be specified to override the default behavior
+
+#### Define copyright owners
+
+The copyright owners for the license can be defined with the `--copyright` flag.
+This will prepend the license header at the top of generated files with a copyright string.
+e.g. *Copyright 2018 The Kubernetes Authors.*
+
+
+```sh
+kubegen versioned --copyright "The Kubernetes Authors."
+```
+
+#### Define the license file
+
+The license file can be overridden from the LICENSE file using the `--license-file` flag.
+This will use the contents from the boilerplate.go.txt file instead of LICENSE.
+If boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments.
+
+
+**Note**: This passes the `--go-header-file` to each of the code generators
+
+```sh
+kubegen versioned --license-file "boilerplate.go.txt"
+```
+
+#### Define the license type
+
+If no license file exists, one may be specified on the using the `--license` flag.
+
+```sh
+kubegen versioned --license "Apache 2.0"
+```
+
+Options:
+
+- "Apache 2.0"
+- "None"
+
+#### Define API groups and versions
+
+The set of APIs code generation is performed for can be explicitly defined instead of
+implicitly determined using the `--api-version` and `--api-group` flags.  When these
+flags are provided, API groups will not be implicitly discovered.  Only API versions
+for groups specified with `--api-group` will be discovered.
+
+```sh
+kubegen versioned --api-version apps/v1 --api-version apps/v1beta1 --api-version extensions/v1beta1
+```
+
+#### Manually specifying the output for the versioned client packages
+
+By default versioned client related libraries are output to the following locations:
+
+- client: `pkg/client/clientset_generated/clientset`
+- informers: `pkg/client/informers_generated/externalversions`
+- listers: `pkg/client/listers_generated`
+
+These locations can be changed using the flags `--client-output`, ``--informer-output` and `--listers-output`.
+This is useful if generating a client for external consumption.
+
+**Note**: These flags also work with `kubegen internal` since that runs generators on versioned types.
+
+```sh
+kubegen versioned --client-output kubernetes --informers-output informers --listers-output listers
+```
+
+#### Running in dry-run
+
+To verify the output without writing actual files use the `--dry-run` flag
+
+```sh
+kubegen versioned --dry-run
+```
+
+**Note**: This passes the `--verify-only` flag to each of the code generators
+
+  
+#### Overriding the searched the api directories
+
+The `--apis-dir` defaults to `pkg/apis` and looks for API groups in that directory.
+
+Looking for APIs outside the default location can be configured using
+the `--apis-dir` flag.  This flag may be provided multiple times to search multiple
+directories.
+
+```sh
+kubegen versioned --apis-dir notpkg/apis --apis-dir pkg/notapis
```

Comment:
If multiple `--apis-dir` and multiple `--api-version` are specified, will the tool try all combinations?

Reply:
> also, they're not under <group>/<version>, they're under <something-vaguely-group-related>/<version>. For example, the types for storage.k8s.io/v1 are in the path storage/v1

The assumption is that <something-vaguely-group-related> is equal for the internal and versioned types. Other than that it must be unique in various generators, i.e. we cannot create a clientset for different groups in a directory `foo` in different apis-dirs.

IMO, we don't need this flexibility of multiple  `--apis-dir`s. Instead, I would prefer to keep it simple and have something more explicit and with less magic like `--internal-apis-dir` and `--versioned-apis-dir`.

Reply:
I think having 1 flag is a simpler interface, than separate flags for internal and versioned api directories.

Reply:
Calling different things with the same name doesn't make it simpler IMO, only leads to magic behind the scenes that requires more knowledge from the user to understand than with two flags.